core/src/main/scala/kafka/log/Log.scala

```diff
@@ -436,8 +466,8 @@ class Log(@volatile var dir: File,
     * @throws KafkaStorageException If the append fails due to an I/O error.
     * @return Information about the appended messages including the first and last offset.
     */
-  def appendAsLeader(records: MemoryRecords, leaderEpoch: Int): LogAppendInfo = {
-    append(records, assignOffsets = true, leaderEpoch)
+  def appendAsLeader(records: MemoryRecords, leaderEpoch: Int, isFromClient: Boolean = true): LogAppendInfo = {
```

Comment:
Yeah, it was a tough call. There's only a handful of use cases for `isFromClient = false`, and the default is the more restrictive option, so I thought it was reasonable.